cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -236,67 +190,6 @@ class ChunkedArraySorter : public TypeVisitor {
   NullPartitionResult* output_;
 };
 
-template <>
-void ChunkedArraySorter::MergeNonNulls<DictionaryType>(
```

Comment:
Here's an example of my intended idea (may not work as is, and may not even compile):
```cpp
class DictionaryMergeSorter {
  template <typename T>
  std::enable_if_t<has_string_view<T>::value || has_c_type<T>::value, Status> Visit(
      const T& type, const SortOrder order, uint64_t* range_begin, uint64_t* range_middle,
      uint64_t* range_end, const std::vector<const Array*>& arrays,
      uint64_t* temp_indices) {
    using ArrayType = typename TypeTraits<T>::ArrayType;

    // This lambda functions should probably be refactored to codegen_internal.h or
    // chunked_internal.h
    auto GetView = [](const Array& array, uint64_t index) {
      const auto& dictionary_array = checked_cast<const DictionaryArray&>(array);
      // GetValueIndex is expensive and can be further optimized
      int64_t value_index = dictionary_array.GetValueIndex(index);
      return checked_cast<const ArrayType&>(*dictionary_array.dictionary())
          .GetView(value_index);
    };

    const ChunkedArrayResolver left_resolver(arrays);
    const ChunkedArrayResolver right_resolver(arrays);
    if (order == SortOrder::Ascending) {
      std::merge(range_begin, range_middle, range_middle, range_end, temp_indices,
                 [&](uint64_t left, uint64_t right) {
                   const auto chunk_left = left_resolver.Resolve<Array>(left);
                   const auto chunk_right = right_resolver.Resolve<Array>(right);
                   auto left_value = GetView(*chunk_left.array, chunk_left.index);
                   auto right_value = GetView(*chunk_right.array, chunk_right.index);
                   return left_value < right_value;
                 });
    } else {
      std::merge(range_begin, range_middle, range_middle, range_end, temp_indices,
                 [&](uint64_t left, uint64_t right) {
                   const auto chunk_left = left_resolver.Resolve<Array>(left);
                   const auto chunk_right = right_resolver.Resolve<Array>(right);
                   auto left_value = GetView(*chunk_left.array, chunk_left.index);
                   auto right_value = GetView(*chunk_right.array, chunk_right.index);
                   return left_value > right_value;
                 });
    }
  }

  Status Visit(const DataType& type, const SortOrder order, uint64_t* range_begin,
               uint64_t* range_middle, uint64_t* range_end,
               const std::vector<const Array*>& arrays, uint64_t* temp_indices) {
    return Status::TypeError("Unsupported type for Dictionary sorting: ",
                             type.ToString());
  }

 public:
  Status Merge(const SortOrder order, uint64_t* range_begin, uint64_t* range_middle,
               uint64_t* range_end, const std::vector<const Array*>& arrays,
               uint64_t* temp_indices) {
    auto dict_value_type =
        checked_cast<const DictionaryArray&>(arrays[0]).dictionary()->type();
    return VisitTypeInline(*dict_value_type, this, range_begin, range_middle, range_end,
                           arrays, temp_indices);
  }
};

template <>
Status ChunkedArraySorter::MergeNonNulls<DictionaryType>(
    uint64_t* range_begin, uint64_t* range_middle, uint64_t* range_end,
    const std::vector<const Array*>& arrays, uint64_t* temp_indices) {
  DictionaryMergeSorter sorter;
  return sorter.Merge(order_, range_begin, range_middle, range_end, arrays, temp_indices);
}
```
In this way there is one type dispatch per chunk, rather than one per element when using a variant. The dispatching can even be moved up further, but I'm not sure how to do it elegantly.